commit 4dd8d833c9350d42528ada0fd65aee41b712f41d
Author: Steve Dickson <steved@redhat.com>
Date: Tue, 20 Jul 2021 17:14:04 -0400
Subject: mount.nfs: Fix the sloppy option processing
 The new mount API broke how the sloppy option is parsed.
 So the option processing needs to be moved up in
 the mount.nfs command.
 
 The option needs to be the first option in the string
 that is passed into the kernel with the -s mount(8)
 and/or the -o sloppy is used.
 
 Commit 92b664ef fixed the process of the -s flag
 and this version fixes the -o sloppy processing
 as well works when libmount-mount is and is not
 enabled plus cleans up the mount options passed
 to the kernel.
 
 Reviewed-and-tested-by: Dave Wysochanski <dwysocha@redhat.com>
 Signed-off-by: Steve Dickson <steved@redhat.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2025302
Origin: upstream, https://git.linux-nfs.org/?p=steved/nfs-utils.git;a=patch;h=4dd8d833c9350d42528ada0fd65aee41b712f41d
Last-Update: 2023-06-29

Index: nfs-utils-1.3.4/utils/mount/nfs.man
===================================================================
--- nfs-utils-1.3.4.orig/utils/mount/nfs.man	2023-06-29 12:14:24.406087991 +1200
+++ nfs-utils-1.3.4/utils/mount/nfs.man	2023-06-29 12:14:24.402087941 +1200
@@ -512,6 +512,13 @@
 and <kernel_soruce>/Documentation/filesystems/caching
 for detail on how to configure the FS-Cache facility.
 Default value is nofsc.
+.TP 1.5i
+.B sloppy
+The
+.B sloppy
+option is an alternative to specifying
+.BR mount.nfs " -s " option.
+
 .SS "Options for NFS versions 2 and 3 only"
 Use these options, along with the options in the above subsection,
 for NFS versions 2 and 3 only.
Index: nfs-utils-1.3.4/utils/mount/stropts.c
===================================================================
--- nfs-utils-1.3.4.orig/utils/mount/stropts.c	2023-06-29 12:14:24.406087991 +1200
+++ nfs-utils-1.3.4/utils/mount/stropts.c	2023-06-29 12:14:24.402087941 +1200
@@ -302,11 +302,19 @@
 
 static int nfs_insert_sloppy_option(struct mount_options *options)
 {
-	if (!sloppy || linux_version_code() < MAKE_VERSION(2, 6, 27))
+	if (linux_version_code() < MAKE_VERSION(2, 6, 27))
 		return 1;
 
-	if (po_insert(options, "sloppy") == PO_FAILED)
-		return 0;
+	if (po_contains(options, "sloppy")) {
+		po_remove_all(options, "sloppy");
+		sloppy++;
+	}
+
+	if (sloppy) {
+		if (po_insert(options, "sloppy") == PO_FAILED)
+			return 0;
+	}
+
 	return 1;
 }
 
